File: 2021/09/09/minimum_difference.py
"""
https://leetcode.com/problems/minimum-absolute-difference-in-bst/
Author: Andrew Jarombek
Date: 9/9/2021
"""

import math
from typing import Optional


class TreeNode:
    def __init__(self, x, left=None, right=None):
        self.val = x
        self.left = left
        self.right = right


class Solution:
    minimum = None
    previous = None

    def get_minimum_difference(self, root: Optional[TreeNode]) -> int:
        self.minimum = math.inf
        self.previous = None
        self.search(root)
        return self.minimum

    def search(self, root: Optional[TreeNode]) -> None:
        if root.left is not None:
            self.search(root.left)

        if self.previous is not None:
            self.minimum = min(self.minimum, abs(root.val - self.previous))

        self.previous = root.val

        if root.right is not None:
            self.search(root.right)


if __name__ == '__main__':
    solution = Solution()
    assert solution.get_minimum_difference(TreeNode(3, TreeNode(2), TreeNode(1))) == 1

File: 2022/09/09/minimumDifference.js
/**
 * https://leetcode.com/problems/detect-capital/
 * @author Andrew Jarombek
 * @since 9/9/2022
 */

function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

const getMinimumDifference = function(root) {
    const result = dfs(root, {prev: null, min: null});
    return result.min;
};

const dfs = function(root, args) {
    if (root.left) {
        args = dfs(root.left, args);
    }

    let {prev, min} = args;

    if (prev !== null) {
        if (min === null) {
            min = root.val - prev;
        } else {
            min = Math.min(min, root.val - prev);
        }
    }

    prev = root.val;
    args = {prev, min};

    if (root.right) {
        args = dfs(root.right, args);
    }

    return args;
}

let result = getMinimumDifference(new TreeNode(1, null, new TreeNode(6, new TreeNode(4), null)));
console.assert(result === 2);
